### PR TITLE
Support Data Source Template Generation

### PR DESCRIPTION
**Pull request by dOrgJelli**
Adding support for datasource templates to the subgraph generation script.

An example of what the input-output of this script now supports:
Input: `src/mappings/DAORegistry/datasource.yaml`
```yml
abis:
  - DAORegistry
entities:
  - DAORegistryContract
eventHandlers:
  - event: Propose(address)
    handler: handlePropose
  - event: Register(address,string)
    handler: handleRegister
  - event: UnRegister(address)
    handler: handleUnRegister
templates:
  - Avatar
  - Controller
  - Reputation
  - DAOToken
```

Output: `subgraph.yaml`
```yml
# ...
- kind: ethereum/contract
    name: DAORegistry
    network: private
    source:
      address: '0x67B5656d60a809915323Bf2C40A8bEF15A152e3e'
      abi: DAORegistry
    mapping:
      kind: ethereum/events
      apiVersion: 0.0.1
      language: wasm/assemblyscript
      file: src/mappings/DAORegistry/mapping.ts
      entities:
        - DAORegistryContract
      abis:
        - name: DAORegistry
          file: ./abis/DAORegistry.json
      eventHandlers:
        - event: Propose(address)
          handler: handlePropose
        - event: 'Register(address,string)'
          handler: handleRegister
        - event: UnRegister(address)
          handler: handleUnRegister
    templates:
      - kind: ethereum/contract
        name: Avatar
        network: private
        source:
          abi: Avatar
        mapping:
          kind: ethereum/events
          apiVersion: 0.0.1
          language: wasm/assemblyscript
          file: src/mappings/Avatar/mapping.ts
          entities:
            - AvatarContract
          abis:
            - name: Avatar
              file: ./abis/Avatar.json
          eventHandlers:
            - event: 'SendEther(uint256,address)'
              handler: handleSendEth
            - event: 'ReceiveEther(address,uint256)'
              handler: handleReceiveEth
      - kind: ethereum/contract
        name: Controller
        network: private
        source:
          abi: Controller
        mapping:
          kind: ethereum/events
          apiVersion: 0.0.1
          language: wasm/assemblyscript
          file: src/mappings/Controller/mapping.ts
          entities:
            - ControllerOrganization
            - ControllerScheme
            - ControllerGlobalConstraint
            - ControllerRegisterScheme
            - ControllerUnregisterScheme
            - ControllerUpgradeController
            - ControllerAddGlobalConstraint
            - ControllerRemoveGlobalConstraint
          abis:
            - name: Controller
              file: ./abis/Controller.json
            - name: Avatar
              file: ./abis/Avatar.json
            - name: Reputation
              file: ./abis/Reputation.json
            - name: DAOToken
              file: ./abis/DAOToken.json
          eventHandlers:
            - event: 'RegisterScheme(address,address)'
              handler: handleRegisterScheme
            - event: 'UnregisterScheme(address,address)'
              handler: handleUnregisterScheme
            - event: 'UpgradeController(address,address)'
              handler: handleUpgradeController
            - event: 'AddGlobalConstraint(address,bytes32,uint8)'
              handler: handleAddGlobalConstraint
            - event: 'RemoveGlobalConstraint(address,uint256,bool)'
              handler: handleRemoveGlobalConstraint
      - kind: ethereum/contract
        name: Reputation
        network: private
        source:
          abi: Reputation
        mapping:
          kind: ethereum/events
          apiVersion: 0.0.1
          language: wasm/assemblyscript
          file: src/mappings/Reputation/mapping.ts
          entities:
            - ReputationContract
            - ReputationHolder
            - ReputationMint
            - ReputationBurn
          abis:
            - name: Reputation
              file: ./abis/Reputation.json
          eventHandlers:
            - event: 'Mint(address,uint256)'
              handler: handleMint
            - event: 'Burn(address,uint256)'
              handler: handleBurn
      - kind: ethereum/contract
        name: DAOToken
        network: private
        source:
          abi: DAOToken
        mapping:
          kind: ethereum/events
          apiVersion: 0.0.1
          language: wasm/assemblyscript
          file: src/mappings/DAOToken/mapping.ts
          entities:
            - TokenHolder
            - TokenTransfer
            - TokenApproval
          abis:
            - name: DAOToken
              file: ./abis/DAOToken.json
          eventHandlers:
            - event: 'Approval(address,address,uint256)'
              handler: handleApproval
            - event: 'Transfer(address,address,uint256)'
              handler: handleTransfer
            - event: 'OwnershipTransferred(address,address)'
              handler: handleOwnershipTransferred
```

### CODE REVIEW

**Comment by dOrgJelli** [comment]
The documentation referenced in the README is currently being PR'd in ATM: https://github.com/graphprotocol/graph-node/pull/910

**Comment by dOrgJelli** [comment]
@orenyodfat this branch is now merged with your changes and should be good to go.

**Comment by orenyodfat** [comment]
how about adding a test for the data source template generation ?

**Comment by dOrgJelli** [comment]
@orenyodfat adding tests now.

**Comment by dOrgJelli** [comment]
Also, please note that adding this update will not change anything with the subgraph's runtime. It simply adds support for the `template:` field in the `mapping/*/datasource.yml` files, so merging this will not break anything.

I've tested manually to make sure the subgraph.yaml that's generated is correctly formatted, but again if none of the `mapping/*/datasource.yml` files are using this new `templates:` field, the subgraph will continue to function as expected with no changes.

**Comment by dOrgJelli** [comment]
It looks like a dependency is breaking the build on all branches (try running tests from master). I think it may be a problem with the postgres container? Looking into this now. @orenyodfat

**Comment by orenyodfat** [comment]
there was an issue with the migration package
please rebase on master
.

> It looks like a dependency is breaking the build on all branches (try running tests from master). I think it may be a problem with the postgres container? Looking into this now. @orenyodfat

there was an issue with the migration package
please rebase on master
.

**Comment by dOrgJelli** [comment]
@orenyodfat apologies for my lag on getting back to you on this today, had some unfortunate technical issues.

> yea, though we need a test to verify that the data source template generation works as expected...
lets have the test have a simple contract which is dynamicly indexed

So I do agree, but I don't currently see an easy way to do this. There are two paths, each with their own downsides:
1. We could add a dynamic datasource to one of the mappings (`src/mappings/*`), but if this is only for testing purposes, I don't think we should add test specific code to the production subgraph. I can't think of a production use-case for a dynamic datasource that would make sense and wouldn't be potentially controversial.
2. We could create tests for the `ops/generate-subgraph.js` script to (1) ensure the generated `subgraph.yaml` that's generated is valid when the `templates:` field is used in the `datasource.yaml` files and (2) ensure the run-time behavior of the mappings works when they try to create these datasources. I've started implementing this, and it's looking to be too much work than it's worth, as we basically get (1) for free from graph-node's CLI when we generate the `subgraph.yaml` (it throws errors if it's invalid), and we'll have (2) covered once we add a production use-case for dynamic datasource to one of the mappings in `src/mappings/*`, and add tests for that. So instead of adding a completely new testing pipeline that will require use to maintain its trickery in the future, I would rather we merge this now, and test for (2) when we add it to the production mappings.

Apologies if this feels like I'm being evasive. I'm certainly on-board with testing everything, in order to test the uncovered ground (run-time functionality) now I feel it'd degrade the maintainability of the project in the long term to add this messy testing pipeline.

Let me know what you think @orenyodfat.

**Comment by dOrgJelli** [comment]
Here's a sample project I'm creating to show the dynamic datasource feature working at run-time: https://github.com/dOrgJelli/dynamic-datasource-test

**Comment by orenyodfat** [comment]
how about adding dao (avatar,reputation,token,controller) as a template data source  ?
we any way need that in production.

**Comment by dOrgJelli** [comment]
@orenyodfat I've gone ahead and added (avatar, reputation, token, controller) to the DAORegistry mapping's datasource.yaml. They are not being created at run-time though as we still haven't reached consensus on how this will best work. I'm thinking actually using these templates will be another PR.

If you'd like to see an example of datasource templates being successfully used at run-time, checkout this sample project I made that does this: https://github.com/dOrgJelli/datasource-template-test

**Comment by orenyodfat** [comment]
how about deploy avatar as part of the test and test that ?

**Comment by dOrgJelli** [comment]
@orenyodfat tests have been added. Please see these files for the changes made:
`src/mappings/DAORegistry/mapping.ts`
`test/{version}/DAORegistry.spec.ts`

**Comment by dOrgJelli** [comment]
working on getting to the bottom of this. Looks like this is causing an issue (not getting the current balance): https://github.com/dOrgTech/subgraph/blob/cefb8af6aabd32be7cd61709d0435bfa0bdbc63a/src/mappings/Avatar/mapping.ts#L46

**Comment by dOrgJelli** [comment]
@orenyodfat this should be good to go now. Let me know if there's anything else I can do to help get this merged.

**Comment by ben-kaufman** [comment]
Hi @dOrgJelli @orenyodfat I see this PR is standing here for a very long time. Is it still relevant?

**Comment by dOrgJelli** [comment]
Yes, this is still relevant for adding new DAOs to the subgraph, and having them show up in Alchemy.

The plan is to:
- Use the DAOtracker contract shown here: https://github.com/daostack/arc/pull/640
- Add DAOtracker to the subgraph
- Immediately after a DAO's avatar, controller, token, and reputation are deployed, it'll be added to he tracker and indexed by the subgraph.

This PR simply adds support for Dynamic Datasources to DAOstack's subgraph. I really wish this had been merged sooner so we didn't have to deal with more merge conflicts :/

**Comment by ben-kaufman** [comment]
Thanks @dOrgJelli , sorry it wasn't merged before. I can take the work of resolving the conflicts and making sure it's good to merge if you want (just respond here to let me know what you prefer).

**Comment by dOrgJelli** [comment]
Hey @ben-kaufman, it's okay I'll go ahead and merge it now.

**Comment by dOrgJelli** [comment]
@ben-kaufman it should be good to go :+1: please let me know if there's anything else I can do to get this merged.

File: src/mappings/DAORegistry/mapping.ts [hunk 1]
```diff
@@ -3,11 +3,20 @@ import 'allocator/arena';
...
 // Import event types from the Token contract ABI
-import { DAORegistry, Propose, Register, UnRegister } from '../../types/DAORegistry/DAORegistry';
+import { Propose, Register, UnRegister } from '../../types/DAORegistry/DAORegistry';
+
+import { Avatar } from '../../types/DAORegistry/templates';
 
 import * as domain from '../../domain';
 
 export function handlePropose(event: Propose): void {
+  // Start tracking the new avatar template if we aren't already
```
Comment (orenyodfat) [inline src/mappings/DAORegistry/mapping.ts]: the purpose of daoRegistery is to index new daos in the dapp(alchemy) layer.
I guess the daotracker(according to the pr description) is the one the subgraph should follow in order to index a new dao .

**Comment by ben-kaufman** [comment]
Hi @dOrgJelli I just tried running this on Rinkeby, it reaches the `Avatar.create(event.params._avatar);` line but does not add the avatar. Any idea why?

File: src/mappings/DAORegistry/mapping.ts (hunk 1 above)
Comment (dOrgJelli) [reply src/mappings/DAORegistry/mapping.ts]: Yes I agree, the reason this was added was to show that the feature was working. This was requested further up in the PR, along with corresponding tests, to validate datasource templates was working. It's essentially stub code to show the functionality. I'm in favor of removing this.

**Comment by dOrgJelli** [comment]
@ben-kaufman can you describe how to repro? Should I run the tests against rinkby?

**Comment by ben-kaufman** [comment]
@dOrgJelli just write this in your `.env`
```
network="rinkeby"
ethereum_node="https://rinkeby.infura.io/v3/69b4a42f71a8446e8b167062e3d5e317"
```
and run:
```
npm run docker:run-rinkeby
npm run deploy
```

**Comment by dOrgJelli** [comment]
@ben-kaufman this ran successfully for me, did you run `npm run test` afterwards and that's where the stall was seen?

**Comment by ben-kaufman** [comment]
@dOrgJelli no, the issue was with running on Rinkeby, not local testing. On Rinkeby, it didn't create a new avatar entity when avatar was proposed to the daoregistry.

**Comment by dOrgJelli** [comment]
@ben-kaufman can you please provide details on how you are reproducing and observing this behavior so I can do so on my end?

**Comment by ben-kaufman** [comment]
Well, just like I wrote with the above instructions, run them and try to query the avatar contracts.

**Comment by dOrgJelli** [comment]
Could you please provide your graphql query.

**Comment by ben-kaufman** [comment]
I use this query:
```
{
  avatarContracts {
    address
    name
  }
}
```

Note a few days ago I proposed an Avatar to the Rinkeby DAORegistry here:
https://rinkeby.etherscan.io/tx/0xb28fc5a036864cf2c87343661b19d0c798c15ff3183e8f9364c8f59c2bb68d4c

However, I don't see it when running the query

*Note: Make sure to empty the `daos/rinkeby` folder before running the deploy command.*

**Comment by dOrgJelli** [comment]
Awesome, thank you for the details @ben-kaufman.

The reason you're not seeing your avatar:
In [`handlePropose`](https://github.com/daostack/subgraph/blob/cfbdb2d4c011fc5fd88ef3c3cbc3f6e890496d2a/src/mappings/DAORegistry/mapping.ts#L17), we're only tracking the new Avatar contract, and this does not mean a new `avatarContract` is added to the database. If we wanted to do this, we could add a lines that call [`getAvatar`](https://github.com/daostack/subgraph/blob/cfbdb2d4c011fc5fd88ef3c3cbc3f6e890496d2a/src/mappings/Avatar/mapping.ts#L13) & [`saveAvatar`](https://github.com/daostack/subgraph/blob/cfbdb2d4c011fc5fd88ef3c3cbc3f6e890496d2a/src/mappings/Avatar/mapping.ts#L27), which will create/init/save a new AvatarContract in the database. This is error prone though because the balance of the new contract has to be set to 0 because we cannot query the current balance of the contract within our mappings.

In order to see the new AvatarContract without doing the above, you need to cause another event to fire from the contract (data source). This is what the [tests are doing](https://github.com/daostack/subgraph/blob/cfbdb2d4c011fc5fd88ef3c3cbc3f6e890496d2a/test/0.0.1-rc.16/DAORegistry.spec.ts#L41) (`transferOwnership` causes an event, which will cause the new Avatar to show up in the database).

**I would prefer not having this code in the DAOregistry mapping at all.**
The code added to the DAOregistry mapping that tracks new avatar contracts was only intended for test purposes and requested by Oren to prove this works. It will be removed when the [DAOtracker is added to Arc](https://github.com/daostack/arc/pull/640). I think we should just remove this code as it's incomplete and doesn't make sense, and might break code that queries the `avatarContracts` expecting a complete DAO to come from it. We have proven the functionality works, and that the subgraph generation works with the new "templates" feature.

**Comment by dOrgJelli** [comment]
Also for added proof of why you're not seeing your avatar:
[created @ block # 5023869](https://rinkeby.etherscan.io/tx/0xb28fc5a036864cf2c87343661b19d0c798c15ff3183e8f9364c8f59c2bb68d4c)
[latest event from contract @ block # 4174484](https://rinkeby.etherscan.io/address/0x72939947e7a1c4ac94bb840e3304b322237ad1a8#events)

**Comment by ben-kaufman** [comment]
@dOrgJelli thanks for the explanation. For me the PR is good, @orenyodfat please review as well.

**Comment by dOrgJelli** [comment]
@orenyodfat @ben-kaufman all changes that affect the run-time behaviour of the subgraph have been removed. This PR only contains changes to the build pipeline, enabling the use of `templates:` in the datasource.yaml files.

**Review by orenyodfat — APPROVED** [review]
👍

**Comment by joeleonjr** [comment]
Hi! I'm a security researcher @TruffleSecurity. I found a valid API key leaked in a comment in this Issue. I highly recommend rotating the API key immediately (not just deleting the comment). Here's a blog post I wrote on remediating these types of vulnerabilities: https://trufflesecurity.com/blog/remediate-leaked-api-keys-with-key-rotation.<br><br>If you can't find the leaked API key, you can use the following [TruffleHog](https://github.com/trufflesecurity/trufflehog) command to automatically scan for it:<br>`trufflehog github --repo https://github.com/<your_username>/<your_reponame> --issue-comments --only-verified`<br><br>If it's not clear where/what the security issue is, please let me know. I'd be happy to help. Please note that the API key might be in the past edit of a comment.<br><br><br><br>(cc @ben-kaufman)

**Pull request merged**